code/datums/ai/ai_controller.dm

```diff
@@ -0,0 +1,753 @@
+RESTRICT_TYPE(/datum/ai_controller)
+
+/**
+ * AI controllers are a datumized form of AI that simulates the input a player
+ * would otherwise give to a atom. What this means is that these datums have
+ * ways of interacting with a specific atom and control it. They posses a
+ * "blackboard" with the information the AI knows and has, and will plan actions
+ * it will try to perform  through multiple modular subtrees with behaviors.
+*/
+/datum/ai_controller
+	/// The atom this controller is controlling.
+	var/atom/pawn
+
+	/**
+	 * This is a list of variables the AI uses and can be mutated by actions.
+	 *
+	 * When an action is performed you pass this list and any relevant keys for
+	 * the variables it can mutate.
+	 *
+	 * DO NOT set values in the blackboard directly, and especially not if
+	 * you're adding a datum reference to this! Use the setters, this is
+	 * important for reference handing.
```

Comment:
You could probably use the spacemandmm var_private/protected on the blackboard to indicate this

Reply:
The blackboard is still expected to be read externally (e.g. `controller.blackboard[BB_BASIC_MOB_CURRENT_TARGET]`), it just can't be directly written to. It should probably be encapsulated into something else at some point but restricting its visibility here wouldn't work for our purposes.